### PR TITLE
More Metrics

### PR DESCRIPTION
**Pull request by mattiecnvr**
* Creates the `ethers-prometheus` crate which provides an Ethers Middlewear which exports metrics via prometheus.
* Extended metrics in `abacus-base` to connect up with the new provider metrics
* Updated `abacus-ethereum` to use the metrics middlewear if the metrics configuration is provided
* Setup other metrics that were previously unused from `abacus-base`
* Updated configs for local testing

Relates to #421

### CODE REVIEW

File: rust/ethers-prometheus/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+            format!("{chain_id}")
+        }
+    } else {
+        "unknown".into()
+    }
+}
+
+/// Convert a u256 scaled integer value into the corresponding f64 value.
+fn u256_as_scaled_f64(value: U256, decimals: u8) -> f64 {
+    value.to_f64_lossy() / (10u64.pow(decimals as u32) as f64)
```
Comment (tkporter) [inline rust/ethers-prometheus/src/lib.rs:499]: whered `to_f64_lossy` move to? for some reason can't find it

**Review by tkporter — APPROVED** [review]
great work! some beautiful rust code

File: rust/abacus-base/src/metrics/core.rs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+            tokio::spawn(async move {
+                warp::serve(
+                    warp::path!("metrics")
+                        .map(move || {
+                            warp::reply::with_header(
+                                self.gather().expect("failed to encode metrics"),
+                                "Content-Type",
+                                // OpenMetrics specs demands "application/openmetrics-text; version=1.0.0; charset=utf-8"
+                                // but the prometheus scraper itself doesn't seem to care?
+                                // try text/plain to make web browsers happy.
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/abacus-base/src/metrics/core.rs:275]: This code was moved, not written by me. So no idea. I will say I was able to access it in the web browser without issue an in the past I have just served the data raw without any info provided and I seem to recall that working. The scraper does not really seem to care much.

File: rust/ethers-prometheus/src/erc20.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+use ethers::prelude::*;
+
+abigen!(Erc20, "$CARGO_MANIFEST_DIR/abis/erc20.abi.json");
+
+// /// Allows us to look up new information on the fly using the ERC20 specification.
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/ethers-prometheus/src/erc20.rs:5]: done

File: rust/ethers-prometheus/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+    /// The wallets to track and identifying info
+    #[serde(default)]
+    pub wallets: HashMap<Address, WalletInfo>,
+
+    /// Contract info for more useful metrics
+    #[serde(default)]
+    pub contracts: HashMap<Address, ContractInfo>,
+}
+
+assert_impl_all!(PrometheusMiddlewareConf: Send, Sync);
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/ethers-prometheus/src/lib.rs:191]: It is a requirement to allow the task thread. This way if the requirements are not met the error is a little more clear.

File: rust/ethers-prometheus/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+                        .map(|n| (a.to_string(), n)),
+                })
+                .unwrap_or_else(|| ("".into(), "unknown".into()));
+
+            m.with(&hashmap! {
+                "chain" => chain_name.as_str(),
+                "contract_name" => contract_name.as_str(),
+                "contract_address" => contract_addr.as_str()
+            })
+            .observe((Instant::now() - start).as_secs_f64())
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/ethers-prometheus/src/lib.rs:287]: I have done some benchmarking before honestly the overhead is negligible, like should be <0.1µs, but even if it weren't negligible the "time to call" should include all overhead as it does not return until after the overhead has been processed.

File: rust/ethers-prometheus/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+}
+
+impl<M: Middleware> PrometheusMiddleware<M> {
+    /// Start the update cycle using tokio. This must be called if you want
+    /// some metrics to be updated automatically. Alternatively you could call update yourself.
+    pub fn start_updating_on_interval(
+        self: &Arc<Self>,
+        period: Duration,
+    ) -> impl Future<Output = ()> + Send {
+        let zelf = Arc::downgrade(self);
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/ethers-prometheus/src/lib.rs:342]: Ah yeah, so this is fun. Basically we are creating a long-running future that does a thing on an interval. We take `self: &Arc<self>` so that we can create a weak pointer, and then stop running when the weak reference ceases to resolve indicating all instances of the Arc were dropped.

We return a Future from a non-async fn because we want have different lifetimes than the arguments (namely self) and we achieve this by making an owned weak reference that we can move into the future. If it were an async fn we would have to have `self: Weak<Self>` which is not currently supported but Arc<Self> is (see custom receivers).

Basically this allows the caller to spin up this thread and then exit because the lifetime is not dependent on the caller.

File: rust/ethers-prometheus/src/lib.rs (hunk 1 above)
Comment (mattiecnvr) [reply rust/ethers-prometheus/src/lib.rs:499]: Had to enable a feature flag for a sub dependency to access it.
https://github.com/paritytech/parity-common/blob/master/primitive-types/src/fp_conversion.rs#L35
